File: Typecast/TCDirectoryEntry.m
//
//  TCDirectoryEntry.m
//  Typecast
//
//  Created by David Schweinsberg on 1/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCDirectoryEntry.h"
#import "TCDataInput.h"

@implementation TCDirectoryEntry

- (id)initWithDataInput:(TCDataInput *)dataInput
{
    self = [super init];
    if (self)
    {
        _tag = [dataInput readInt];
        _checksum = [dataInput readInt];
        _offset = [dataInput readInt];
        _length = [dataInput readInt];
   }
    return self;
}

- (NSString *)tagAsString
{
    return [NSString stringWithFormat:
            @"%c%c%c%c",
            (char)(_tag >> 24),
            (char)(_tag >> 16),
            (char)(_tag >> 8),
            (char)_tag];
}

- (NSString *)description
{
    return [NSString stringWithFormat:
            @"'%@' - chksm = 0x%x, off = 0x%x, len = %d",
            [self tagAsString],
            _checksum,
            _offset,
            _length];
}

#pragma mark - NSCopying Methods

- (id)copyWithZone:(NSZone *)zone
{
    TCDirectoryEntry *entryCopy = [[TCDirectoryEntry allocWithZone:zone] init];
    entryCopy.tag = _tag;
    entryCopy.checksum = _checksum;
    entryCopy.offset = _offset;
    entryCopy.length = _length;
    return entryCopy;
}

@end

File: Typecast/TCDirectoryEntry.swift
//
//  TCDirectoryEntry.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/31/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

struct TCDirectoryEntry: CustomStringConvertible {
  let tag: UInt32
  let checksum: UInt32
  let offset: Int
  let length: Int
  var tagAsString: String {
    get {
      return String(format: "%c%c%c%c",
                    CChar(truncatingIfNeeded: tag >> 24),
                    CChar(truncatingIfNeeded: tag >> 16),
                    CChar(truncatingIfNeeded: tag >> 8),
                    CChar(truncatingIfNeeded: tag))
    }
  }

  init(dataInput: TCDataInput) {
    tag = dataInput.readUInt32()
    checksum = dataInput.readUInt32()
    offset = Int(dataInput.readUInt32())
    length = Int(dataInput.readUInt32())
  }

  init(tag: UInt32, checksum: UInt32, offset: Int, length: Int) {
    self.tag = tag
    self.checksum = checksum
    self.offset = offset
    self.length = length
  }

  var description: String {
    get {
      return String(format:"'%@' - chksm = 0x%x, off = 0x%x, len = %d",
                    tagAsString,
                    checksum,
                    offset,
                    length)
    }
  }
}
